### PR TITLE
feat(new): Added Azure.AppService.MigrateNodeJs

### PR DESCRIPTION
**Pull request by BenjaminEngeset**
## PR Summary

Working toward #2879

Added Azure.AppService.MigrateNodeJs.

## PR Checklist

- [x] PR has a meaningful title
- [ ] Summarized changes
- [ ] Change is not breaking
- [ ] This PR is ready to merge and is not **Work in Progress**
- **Rule changes**
  - [ ] Unit tests created/ updated
  - [ ] Rule documentation created/ updated
  - [ ] Link to a filed issue
  - [ ] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section
- **Other code changes**
  - [ ] Unit tests created/ updated
  - [ ] Link to a filed issue
  - [ ] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

**Comment by BenjaminEngeset** [comment]
Hi @BernieWhite,

Could you please provide some feedback on the work-in-progress code, specifically regarding the rule design/body?
I am interested in your thoughts.

File: src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1 [hunk 1]
```diff
@@ -155,19 +155,68 @@ Rule 'Azure.AppService.WebProbePath' -Ref 'AZR-000080' -With 'Azure.AppService.I
...
+                'FtpsOnly'
+                'Disabled'
+            ));
+    }
+}
+
+# Synopsis: Use supported Node.js versions for the runtime of applications.
+Rule 'Azure.AppService.MigrateNodeJs' -Ref 'AZR-000425' -Type 'Microsoft.Web/sites', 'Microsoft.Web/sites/config', 'Microsoft.Web/sites/slots', 'Microsoft.Web/sites/slots/config' -Tag @{ release = 'GA'; ruleSet = '2024_06'; 'Azure.WAF/pillar' = 'Reliability'; } {
+    
+    [SemVer[]]$versions = @(
```
Comment (BernieWhite) [inline src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1:177]: `SemVer` is a PowerShell 7.x specific type. It doesn't work on Windows PowerShell.

Until we drop support for Windows PowerShell we need to keep compatibility.

But `$Assert.Version` supports parsing semver from a string, so I don't think there is a reason to use to type specifically.

File: src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1 [hunk 2]
```diff
@@ -155,19 +155,68 @@ Rule 'Azure.AppService.WebProbePath' -Ref 'AZR-000080' -With 'Azure.AppService.I
...
+        GetSubResources -ResourceType 'Microsoft.Web/sites/slots' |
+        ForEach-Object { $_.properties.siteConfig.appSettings.WEBSITE_NODE_DEFAULT_VERSION }
+        # App Service on Windows.
+        GetSubResources -ResourceType 'Microsoft.Web/sites/config', 'Microsoft.Web/sites/slots/config' |
+        Where-Object name -eq 'appsettings' |
+        ForEach-Object { $_.properties.WEBSITE_NODE_DEFAULT_VERSION }
+        # App Service on Windows.
+        GetSubResources -ResourceType 'Microsoft.Web/sites/config', 'Microsoft.Web/sites/slots/config' |
+        Where-Object name -eq 'web' |
+        ForEach-Object { $_.properties.appSettings.WEBSITE_NODE_DEFAULT_VERSION }
```
Comment (BernieWhite) [inline src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1:206]: Good job identifying all the possible cases.

This should be functionally fine, but if we can consider breaking it down similar to `GetWebSiteConfig` to reduce some of the complexity.

**Review by BernieWhite — COMMENTED** [review]
Thanks @BenjaminEngeset. See comments below.

File: src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1 (hunk 2 above)
Comment (BenjaminEngeset) [reply src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1:206]: Thanks @BernieWhite. Where is source code for `GetWebSiteConfig`? I have tried to search within this repository, but not able to find it.

File: src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1 (hunk 1 above)
Comment (BenjaminEngeset) [reply src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1:177]: Good news! Thanks.

File: src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1 (hunk 2 above)
Comment (BenjaminEngeset) [reply src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1:206]: When you have the time @BernieWhite, perhaps you didn't see this.

Comment (BernieWhite) [reply src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1:206]: @BenjaminEngeset https://github.com/Azure/PSRule.Rules.Azure/blob/1e28104a65a3de731cc43557cc9ed17b24412c9c/src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1#L201-L210

Comment (BenjaminEngeset) [reply src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1:206]: Thanks @BernieWhite.

**Comment by BenjaminEngeset** [comment]
Hi again @BernieWhite. 

Can you give me some feedback?

I'm still missing the tests, but wanted to check in with you.

File: docs/en/rules/Azure.AppService.NodeJsVersion.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+- For App Service apps on Windows-based plans:
+  - Set the `properties.siteConfig.appSettings.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
+  - It is also possible to set this for sub-resources by:
+    - If the type is `Microsoft.Web/sites/slots`, set the `properties.siteConfig.appSettings.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
+    - If the type is `Microsoft.Web/sites/config` and the `name` property is set to `appsettings`, set the `properties.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
+    - If the type is `Microsoft.Web/sites/slots/config` and the `name` property is set to `appsettings`, set the `properties.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
+    - If the type is `Microsoft.Web/sites/config` and the `name` property is set to `web`, set the `properties.appSettings.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
+    - If the type is `Microsoft.Web/sites/slots/config` and the `name` property is set to `web`, set the `properties.appSettings.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
+
+For example for Linux:
```
Comment (BernieWhite) [inline docs/en/rules/Azure.AppService.NodeJsVersion.md:48]:
```suggestion
For example (Linux app):
```

File: docs/en/rules/Azure.AppService.NodeJsVersion.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+      "linuxFxVersion": "NODE|20-lts"
+    }
+  },
+  "dependsOn": [
+    "[resourceId('Microsoft.Web/serverfarms', parameters('planName'))]"
+  ]
+}
+```
+
+For example for Windows:
```
Comment (BernieWhite) [inline docs/en/rules/Azure.AppService.NodeJsVersion.md:74]:
```suggestion
For example (Windows app):
```

File: docs/en/rules/Azure.AppService.NodeJsVersion.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    "type": "SystemAssigned"
+  },
+  "kind": "web",
+  "properties": {
+    "serverFarmId": "[resourceId('Microsoft.Web/serverfarms', parameters('planName'))]",
+    "httpsOnly": true,
+    "siteConfig": {
+      "minTlsVersion": "1.2",
+      "linuxFxVersion": "~20"
+    }
```
Comment (BernieWhite) [inline docs/en/rules/Azure.AppService.NodeJsVersion.md:92]:
```suggestion
    "siteConfig": {
      "minTlsVersion": "1.2",
      "appSettings": [
        {
          "name": "WEBSITE_NODE_DEFAULT_VERSION",
          "value": "~20"
        }
      ]
    }
```

File: docs/en/rules/Azure.AppService.NodeJsVersion.md [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+- For App Service apps on Windows-based plans:
+  - Set the `properties.siteConfig.appSettings.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
+  - It is also possible to set this for sub-resources by:
+    - If the type is `Microsoft.Web/sites/slots`, set the `properties.siteConfig.appSettings.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
+    - If the type is `Microsoft.Web/sites/config` and the `name` property is set to `appsettings`, set the `properties.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
+    - If the type is `Microsoft.Web/sites/slots/config` and the `name` property is set to `appsettings`, set the `properties.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
+    - If the type is `Microsoft.Web/sites/config` and the `name` property is set to `web`, set the `properties.appSettings.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
+    - If the type is `Microsoft.Web/sites/slots/config` and the `name` property is set to `web`, set the `properties.appSettings.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
```
Comment (BernieWhite) [inline docs/en/rules/Azure.AppService.NodeJsVersion.md:117]: I think we can just simplify this to:

```suggestion
- For Linux-based apps and slots:
  - Set the `properties.siteConfig.linuxFxVersion` property to `NODE|20-lts`.
- For Windows-based apps and slots:
  - Add an app setting within `properties.siteConfig.appSettings` by creating an object with the `name` and `value` properties.
  - Set the `name` property to `WEBSITE_NODE_DEFAULT_VERSION`.
  - Set the `value` property to `~20`.

In addition to setting the `properties.siteConfig` property, you can also use a sub-resource.
```

File: docs/en/rules/Azure.AppService.NodeJsVersion.md [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    httpsOnly: true
+    siteConfig: {
+      minTlsVersion: '1.2'
+      linuxFxVersion: 'NODE|20-lts'
+    }
+  }
+}
+```
+
+For example for Windows:
```
Comment (BernieWhite) [inline docs/en/rules/Azure.AppService.NodeJsVersion.md:140]:
```suggestion
For example (Windows app):
```

File: docs/en/rules/Azure.AppService.NodeJsVersion.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+- For App Service apps on Windows-based plans:
+  - Set the `properties.siteConfig.appSettings.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
+  - It is also possible to set this for sub-resources by:
+    - If the type is `Microsoft.Web/sites/slots`, set the `properties.siteConfig.appSettings.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
+    - If the type is `Microsoft.Web/sites/config` and the `name` property is set to `appsettings`, set the `properties.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
+    - If the type is `Microsoft.Web/sites/slots/config` and the `name` property is set to `appsettings`, set the `properties.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
+    - If the type is `Microsoft.Web/sites/config` and the `name` property is set to `web`, set the `properties.appSettings.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
+    - If the type is `Microsoft.Web/sites/slots/config` and the `name` property is set to `web`, set the `properties.appSettings.WEBSITE_NODE_DEFAULT_VERSION` property to `~20`.
+
+For example for Linux:
```
Comment (BernieWhite) [inline docs/en/rules/Azure.AppService.NodeJsVersion.md:119]:
```suggestion
For example (Linux app):
```

File: docs/en/rules/Azure.AppService.NodeJsVersion.md [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    type: 'SystemAssigned'
+  }
+  kind: 'web'
+  properties: {
+    serverFarmId: plan.id
+    httpsOnly: true
+    siteConfig: {
+      minTlsVersion: '1.2'
+      linuxFxVersion: '~20'
+    }
```
Comment (BernieWhite) [inline docs/en/rules/Azure.AppService.NodeJsVersion.md:156]:
```suggestion
    siteConfig: {
      minTlsVersion: '1.2'
      appSettings: [  
        {  
          name: 'WEBSITE_NODE_DEFAULT_VERSION'
          value: '~20
        }  
      ]  
    }
```

File: src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1 [hunk 10]
```diff
@@ -155,19 +155,37 @@ Rule 'Azure.AppService.WebProbePath' -Ref 'AZR-000080' -With 'Azure.AppService.I
...
-            'Disabled'
-        ));
+                'FtpsOnly'
+                'Disabled'
+            ));
+    }
+}
+
+# Synopsis: Configure applications to use supported Node.js runtime versions.
+Rule 'Azure.AppService.NodeJsVersion' -Ref 'AZR-000428' -Type 'Microsoft.Web/sites', 'Microsoft.Web/sites/config', 'Microsoft.Web/sites/slots', 'Microsoft.Web/sites/slots/config' -Tag @{ release = 'GA'; ruleSet = '2024_06'; 'Azure.WAF/pillar' = 'Reliability'; } {
```
Comment (BernieWhite) [inline src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1:175]:
```suggestion
Rule 'Azure.AppService.NodeJsVersion' -Ref 'AZR-000428' -Type 'Microsoft.Web/sites', 'Microsoft.Web/sites/config', 'Microsoft.Web/sites/slots', 'Microsoft.Web/sites/slots/config' -Tag @{ release = 'GA'; ruleSet = '2024_06'; 'Azure.WAF/pillar' = 'Security'; } {
```

File: src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1 [hunk 11]
```diff
@@ -203,10 +221,59 @@ function global:GetWebSiteConfig {
...
+        # App Service on Linux.
+        GetSubResources -ResourceType 'Microsoft.Web/sites/slots' | 
+        ForEach-Object { $_.properties.siteConfig.linuxFxVersion | Where-Object { $_ -like 'NODE|*' -and $_ -ne 'NODE|lts' } }
+        # App Service on Linux.
+        GetSubResources -ResourceType 'Microsoft.Web/sites/config', 'Microsoft.Web/sites/slots/config' |
+        Where-Object name -eq 'web' |
+        ForEach-Object { $_.properties.linuxFxVersion | Where-Object { $_ -like 'NODE|*' -and $_ -ne 'NODE|lts' } }
+        
+        # App Service on Windows. Works for when main object equals Microsoft.Web/sites or Microsoft.Web/sites/slots
+        $TargetObject.properties.siteConfig.appSettings.WEBSITE_NODE_DEFAULT_VERSION
```
Comment (BernieWhite) [inline src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1:259]: When accessed under the `properties.siteConfig.appSettings` this should be an array, with an name/ value object. See documentation example.

File: src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1 [hunk 12]
```diff
@@ -203,10 +221,59 @@ function global:GetWebSiteConfig {
...
+        # App Service on Windows. Works for when main object equals Microsoft.Web/sites/config 'appsettings' or Microsoft.Web/sites/slots/config 'appsettings'
+        $TargetObject.properties.WEBSITE_NODE_DEFAULT_VERSION
+        # App Service on Windows. Works for when main object equals Microsoft.Web/sites/config 'web' or Microsoft.Web/sites/slots/config 'web'
+        $TargetObject.properties.appSettings.WEBSITE_NODE_DEFAULT_VERSION
+        # App Service on Windows.
+        GetSubResources -ResourceType 'Microsoft.Web/sites/slots' |
+        ForEach-Object { $_.properties.siteConfig.appSettings.WEBSITE_NODE_DEFAULT_VERSION }
```
Comment (BernieWhite) [inline src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1:266]: When accessed under the `properties.siteConfig.appSettings` this should be an array, with an name/ value object. See documentation example.

File: src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1 [hunk 13]
```diff
@@ -203,10 +221,59 @@ function global:GetWebSiteConfig {
...
+        # App Service on Windows.
+        GetSubResources -ResourceType 'Microsoft.Web/sites/config', 'Microsoft.Web/sites/slots/config' |
+        Where-Object name -eq 'appsettings' |
+        ForEach-Object { $_.properties.WEBSITE_NODE_DEFAULT_VERSION }
+        # App Service on Windows.
+        GetSubResources -ResourceType 'Microsoft.Web/sites/config', 'Microsoft.Web/sites/slots/config' |
+        Where-Object name -eq 'web' |
+        ForEach-Object { $_.properties.appSettings.WEBSITE_NODE_DEFAULT_VERSION }
```
Comment (BernieWhite) [inline src/PSRule.Rules.Azure/rules/Azure.AppService.Rule.ps1:274]: When accessed under the `properties.appSettings` this should be an array, with an name/ value object. See documentation example.

**Review by BernieWhite — CHANGES_REQUESTED** [review]
Hi @BenjaminEngeset, great progress.

From some additional research, if you are accessing app settings via the `appsettings` named config then the properties can be set via `{ }`, otherwise they should be:

```
appSettings: [  
  {  
    name: 'WEBSITE_NODE_DEFAULT_VERSION'  
    value: '~20  
  }  
] 
```

**Comment by BenjaminEngeset** [comment]
> Hi @BenjaminEngeset, great progress.
> 
> From some additional research, if you are accessing app settings via the `appsettings` named config then the properties can be set via `{ }`, otherwise they should be:
> 
> ```
> appSettings: [  
>   {  
>     name: 'WEBSITE_NODE_DEFAULT_VERSION'  
>     value: '~20  
>   }  
> ] 
> ```

Great catch! I appreciate you identifying this. I will add in logic for this.

**Comment by BenjaminEngeset** [comment]
So @BernieWhite, what do you think about the rule now?

I have added test scenarios for several cases, but I have not added cases for sub-resources configured within the parent resource. If you feel that is needed, let me know.

**Comment by BernieWhite** [comment]
> So @BernieWhite, what do you think about the rule now?
> 
> I have added test scenarios for several cases, but I have not added cases for sub-resources configured within the parent resource. If you feel that is needed, let me know.

Thanks @BenjaminEngeset Looks look. Yes let's add at least one good and bad case for `Microsoft.Web/sites` sub-resources.

**Comment by BenjaminEngeset** [comment]
I have added a good and bad test case now @BernieWhite.

Following logic:
- In the first case the app setting will override what was set within the parent resource.
- In the second case the app setting will override what was set within the parent resource, but it will set what we expect as good.

**Review by BernieWhite — APPROVED** [review]
The tests looks great. Great work @BenjaminEngeset. All good to merge.

**Pull request merged**